src/layout.cpp

```diff
@@ -0,0 +1,476 @@
+    nodes[i].pred.clear();
+  }
+
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      nodes[nodes[i].succ[j]].pred.push_back(i);
+    }
+  }
+}
+
+void degReinit(std::vector<TreeNode> &nodes) {
+  for (int i = 0; i < nodes.size(); i++) {
+    nodes[i].degP = 0;
+    nodes[i].degM = 0;
+  }
+
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      nodes[i].degP++;
+      nodes[nodes[i].succ[j]].degM++;
+    }
+  }
+}
+
+void heapify(std::vector<TreeNode> &nodes, int n, int i) {
+  int largest = i;
+  int l = 2 * i + 1;
+  int r = 2 * i + 2;
+
+  if (l < n && nodes[l].id > nodes[largest].id) {
+    largest = l;
+  }
+
+  if (r < n && nodes[r].id > nodes[largest].id) {
+    largest = r;
+  }
+
+  if (largest != i) {
+    std::swap(nodes[i], nodes[largest]);
+    heapify(nodes, n, largest);
+  }
+}
+
+void idHeapSort(std::vector<TreeNode> &nodes) {
+  int n = 0;
+  for (int i = 0; i < nodes.size(); i++) {
+    n++;
+  }
+
+  for (int i = n / 2 - 1; i >= 0; i--) {
+    heapify(nodes, n, i);
+  }
+
+  for (int i = n - 1; i >= 0; i--) {
+    std::swap(nodes[0], nodes[i]);
+    heapify(nodes, i, 0);
+  }
+}
+
+bool algorithmDFS(TreeNode::nodeId id, std::vector<TreeNode> &nodes,
+                  std::vector<bool> &usedId) {
+  usedId[id] = true;
+
+  for (int i = 0; i < nodes[id].succ.size(); i++) {
+    if (usedId[nodes[id].succ[i]]) {
+      return true;
+    } else {
+      bool exist = algorithmDFS(nodes[id].succ[i], nodes, usedId);
+      if (exist) {
+        return true;
+      }
+    }
+  }
+
+  usedId[id] = false;
+  return false;
+}
+
+bool cycleExistsDFS(std::vector<TreeNode> &nodes) {
+  degReinit(nodes);
+
+  std::vector<bool> usedId = {};
+  usedId.resize(nodes.size(), false);
+
+  for (int i = 0; i < nodes.size(); i++) {
+    if (nodes[i].degM == 0) {
+      bool exist = algorithmDFS(nodes[i].id, nodes, usedId);
+      if (exist) {
+        return true;
+      }
+    }
+  }
+
+  return algorithmDFS(nodes[0].id, nodes, usedId);
+}
+
+struct EdgeCount {
+  int degP = 0;
+  int degM = 0;
+};
+
+void removeEdgeCount(int i, std::vector<EdgeCount> &edgeCounts,
+                     std::vector<TreeNode> &nodes) {
+  for (int j = 0; j < nodes[i].succ.size(); j++) {
+    edgeCounts[nodes[i].succ[j]].degM--;
+  }
+  for (int j = 0; j < nodes[i].pred.size(); j++) {
+    edgeCounts[nodes[i].pred[j]].degP--;
+  }
+
+  edgeCounts[i].degM = -1;
+  edgeCounts[i].degP = -1;
+}
+
+void greedyFAS(
+    std::vector<TreeNode> &nodes,
+    std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> &deletedEdges) {
+  degReinit(nodes);
+  predReinit(nodes);
+  std::vector<TreeNode::nodeId> s1 = {}, s2 = {};
+  std::vector<EdgeCount> edgeCounts = {};
+  for (int i = 0; i < nodes.size(); i++) {
+    EdgeCount edgeCount = {};
+    edgeCount.degP = nodes[i].degP;
+    edgeCount.degM = nodes[i].degM;
+    edgeCounts.push_back(edgeCount);
+  }
```

Comment:
Do `vector::reserve` when the size is known beforehand.
```suggestion
  std::vector<EdgeCount> edgeCounts;
  edgeCounts.reserve(nodes.size());
  for (const TreeNode &node: nodes) {
    edgeCounts.push_back({ node.degP, node.degM });
  }
```